homeassistant/components/icloud/__init__.py

```diff
@@ -1 +1,629 @@
+        """Call the update function of an iCloud account."""
+        accountname = service.data.get(ATTR_ACCOUNTNAME)
+
+        if accountname is None:
+            for accountname, account in hass.data[DATA_ICLOUD].items():
+                account.keep_alive(utcnow())
+        else:
+            accountname = slugify(accountname.partition('@')[0])
+            hass.data[DATA_ICLOUD][accountname].keep_alive(utcnow())
+    hass.services.register(DOMAIN, SERVICE_ICLOUD_UPDATE, update,
+                           schema=SERVICE_SCHEMA)
+
+    async def reset_account(service):
+        """Reset an iCloud account."""
+        accountname = service.data.get(ATTR_ACCOUNTNAME)
+
+        if accountname is None:
+            for accountname, account in hass.data[DATA_ICLOUD].items():
+                account.reset_account()
+        else:
+            accountname = slugify(accountname.partition('@')[0])
+            hass.data[DATA_ICLOUD][accountname].reset_account()
+
+    hass.services.register(DOMAIN, SERVICE_ICLOUD_RESET,
+                           reset_account, schema=SERVICE_SCHEMA)
+
+    def setup_icloud(icloud_config):
+        """Set up an iCloud account."""
+        _LOGGER.debug("Logging into iCloud...")
+
+        username = icloud_config.get(CONF_USERNAME)
+        password = icloud_config.get(CONF_PASSWORD)
+        account_name = icloud_config.get(CONF_ACCOUNTNAME,
+                                         slugify(username.partition('@')[0]))
+        max_interval = icloud_config.get(CONF_MAX_INTERVAL)
+        gps_accuracy_threshold = icloud_config.get(CONF_GPS_ACCURACY_THRESHOLD)
+
+        account = IcloudAccount(hass, username, password, account_name,
+                                max_interval, gps_accuracy_threshold)
+
+        if account.api is not None:
+            hass.data[DATA_ICLOUD][account.name] = account
+
+        else:
+            _LOGGER.error("No iCloud data added for account=%s", account_name)
+            return False
+
+        for component in ICLOUD_COMPONENTS:
+            load_platform(hass, component, DOMAIN, {}, icloud_config)
+
+    hass.data[DATA_ICLOUD] = {}
+    for icloud_config in config[DOMAIN]:
+        setup_icloud(icloud_config)
+
+    return True
+
+
+class IcloudAccount():
+    """Representation of an iCloud account."""
+
+    def __init__(self, hass, username, password, accountname, max_interval,
+                 gps_accuracy_threshold):
+        """Initialize an iCloud account."""
+        self._hass = hass
+        self.username = username
+        self.__password = password
+        self._accountname = accountname
+        self._max_interval = max_interval
+        self._gps_accuracy_threshold = gps_accuracy_threshold
+
+        self.api = None
+        self.account_owner_fullname = None
+        self.family_members_fullname = {}
+        self.devices = {}
+
+        self.__trusted_device = None
+        self.__verification_code = None
+
+        self.reset_account()
```

Comment:
We don't want side effects in init method. Please move this call outside the class or into a dedicated setup method.

Reply:
Okey, move it to the `setup_icloud` method.

Reply:
For my knowledge: how it can make side effects ?

Reply:
Here's the definition:
https://en.m.wikipedia.org/wiki/Side_effect_(computer_science)

In this case we created a directory and made calls to an api.

Reply:
That work that should be done in a ConfigFlow, and will clean the init a lot !